src/Qowaiv/Security/Cryptography/Secret.cs

```diff
@@ -0,0 +1,63 @@
+﻿using Qowaiv.Diagnostics;
+using Qowaiv.Hashing;
+using System;
+using System.Diagnostics;
+using System.Diagnostics.Contracts;
+
+namespace Qowaiv.Security.Cryptography
+{
+    /// <summary>Represents a secret.</summary>
+    [DebuggerDisplay("{DebuggerDisplay}")]
+    public readonly struct Secret : IEquatable<Secret>
+    {
+        /// <summary>Represents an empty/not set secret.</summary>
+        public static readonly Secret Empty;
+
+        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
+        private readonly string m_Value;
+
+        private Secret(string value) => m_Value = value;
+
+        /// <summary>Returns true if the secret is empty/not set.</summary>
+        [Pure]
+        public bool IsEmpty() => m_Value is null;
+
+        /// <summary>Gets a string representing this secret.</summary>
+        [Pure]
+        public string Value() => m_Value ?? string.Empty;
+
+        /// <inheritdoc />
+        [Pure]
+        public override bool Equals(object obj) => obj is Secret other && Equals(other);
+
+        /// <inheritdoc />
+        [Pure]
+        public bool Equals(Secret other) => m_Value == other.m_Value;
+
+        /// <inheritdoc />
+        [Pure]
+        public override int GetHashCode() => Hash.NotSupportedBy<Secret>();
+
+        /// <inheritdoc />
+        [Pure]
+        public override string ToString() => m_Value is null ? string.Empty : "???";
```

Comment:
Well, its a value type that is nullable by design. If it was a reference type, it would also be different. But you would opt for "???" with `Empty` as well?

Reply:
Probably, but it's more a gut feeling so as to force the user to make his intent clear while at the same time making the behavior less surprising when viewed from the outside. I guess this also intersects with my other comment of whether a secret can be empty and whether this would be a failure case. The question itself may indicate that the answer may actually depend on the particular use-case that you are implementing, which may lead to the question whether this should actually be a responsibility of the `Secret` class.

Also, depending on the use-case, should `Secret` perhaps be generalized to allow for other types of secrets such as byte arrays?

Reply:
There is already a type based on byte arrays. The point is, that you always can create a struct with the following: `var secret = default(Secret);`, so as a consequence of that, we *MUST* have a valid state for a underlying value having a default value, `(string)null` in this case.

Reply:
> There is already a type based on byte arrays.

👍

>  The point is, that you always can create a struct with the following: var secret = default(Secret);, so as a consequence of that, we MUST have a valid state for a underlying value having a default value, (string)null in this case.

Fair enough; I do think this leaves some small room for accidental misuse of the type. But perhaps the presence of `IsEmpty()` is enough documentation to alleviate this in practice.